### PR TITLE
Add Quark reexport

### PR DESCRIPTION
**Pull request by EPashkin**

### CODE REVIEW

**Comment by gkoz** [comment]
I'm not convinced exposing quarks rather than implementing translations between `str`/`String` and `GQuark` is the way to go.

**Comment by EPashkin** [comment]
You right, quarks is special case of alias.
I selected wrong alias for testing.
I will try add proxy (with cache) and conversion in functions body from `&str`

**Comment by EPashkin** [comment]
On second though cache unneeded as glib already use hash table.

**Comment by EPashkin** [comment]
May be better add Quark to Fundamental?

**Comment by gkoz** [comment]
A cache would certainly save some CPU time due to avoiding unnecessary `CString` copying but it sounds like a premature optimization.

As far as I remember, the point of `Fundamental` was to cover the types, that don't have proper gir definitions (`GType` might be somewhat of an exception but it's not referred to properly outside of `GObject`).

**Comment by EPashkin** [comment]
Build fixed.
I started to think that this branch must first merged to main repo and after this replaced with merging with master here (or just closed).

**Comment by gkoz** [comment]
What I had in mind was hiding the existence of quarks altogether but a more explicit approach is possible too, we could use `T: Into<Quark>` in the high level signatures to allow passing either strings or quarks.

`From<u32>` seems dubious because it allows creating invalid quarks and would anyone want to convert to and from `u32` anyway? `get_string` would probably work better as `impl Display`, if we don't let the user create invalid quarks, they shouldn't have a way to obtain them.

It's fine by me to introduce this in this branch and merge into master when it's actually needed there.

**Comment by EPashkin** [comment]
Ready to merge

File: src/quark.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+impl Display for Quark {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        let ptr = unsafe {
+            glib_ffi::g_quark_to_string(self.0)
+        };
+        if ptr.is_null() {
+            Ok(())
+        } else {
+            String::from_utf8_lossy(unsafe { CStr::from_ptr(ptr).to_bytes() })
+                .fmt(f)
```
Comment (EPashkin) [inline src/quark.rs]: Fails only 86x standard, may be just accident

File: src/quark.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+}
+
+impl Deref for Quark {
+    type Target = u32;
```
Comment (gkoz) [inline src/quark.rs]: It's somewhat redundant anyway because `GQuark` is `u32` so `to_glib` achieves exactly the same.

**Comment by EPashkin** [comment]
Removed test and `Deref`

File: src/quark.rs (hunk 2 above)
Comment (gkoz) [inline src/quark.rs]: Oh this is embarrassing. `from_glib` then actually allows to create invalid quarks easily so there goes the assumption that they're valid. We need to check if it's safe to pass them into the library.

**Comment by gkoz** [comment]
Copying into the top level for the record:

> Oh this is embarrassing. `from_glib` then actually allows to create invalid quarks easily so there goes the assumption that they're valid. We need to check if it's safe to pass them into the library.

File: src/quark.rs (hunk 2 above)
Comment (EPashkin) [inline src/quark.rs]: If it not valid, `g_quark_to_string` just return NULL

**Comment by EPashkin** [comment]
Also
> If it not valid, g_quark_to_string just return NULL. So IMHO it safe

**Comment by gkoz** [comment]
It's notable that they always make sure the quark is non-zero and an invalid one is the same as `0` as far as `g_quark_to_string` is concerned. I've found [one case](https://github.com/GNOME/glib/blob/master/gio/gsettings.c#L381-L383) where this comes down to [g_return_val_if_fail](https://developer.gnome.org/glib/stable/glib-Warnings-and-Assertions.html#g-return-val-if-fail) and relying on that is not recommended. That particular function might not ever be reachable from Rust but it demonstrates the assumption that non-zero quarks are valid.

On the other hand the `quark != 0` checks are also under the same `g_return_val_if_fail` so the problem is even bigger.

This pulls strongly toward using direct translations and never exposing the existence of quarks.

**Comment by EPashkin** [comment]
Good reasoning.
So example better changed to next?
```
///! Quark can lead to undefined behavior in gtk\glib code build with G_DISABLE_CHECKS
///! (see https://github.com/GNOME/glib/blob/master/gio/gsettings.c#L381-L383),
///! so it better not used in public interfaces.

pub fn iter_has_qclass(&self, pos: i32, qname: &str) -> bool {
    unsafe {
        from_glib(ffi::gtk_widget_path_iter_has_qclass(self.to_glib_none().0, pos, glib::Quark::from(qname).to_glib()))
    }
}
```

**Comment by gkoz** [comment]
When variants taking `char*` are available (e.g. `_iter_has_class`) the quark variants could just be ignored.

**Comment by EPashkin** [comment]
Then i add `ignored_functions` to object config.
And remove example here.

**Comment by gkoz** [comment]
I guess I'd put this commit in a branch somewhere but as long as nothing needs quarks and because of safety concerns adding it to `master` is for some other time.

**Comment by EPashkin** [comment]
Ok.
 I update https://github.com/gkoz/gir/pull/162 if find other alias that can be used as test

**Pull request closed without merging**